cspann: move vectors from sibling partitions during split

andy-kimball opened the pull request:
#### cspann: move vectors from sibling partitions during split

When a partition is split, the left and right target partitions are assigned
new centroids. This makes it possible for vectors in other partitions at the
same level to now be closer to one of those new centroids than they are to
their own centroid. In that case, we need to move those vectors to whichever
target partition is now closer.

#### cspann: enhance SearchSet

Add support in SearchSet for excluding partitions during search. Any
vectors in these partitions will not be added to the set. Also add
an option that includes the distance of each vector from its centroid
in search results. These options will be used by the split operation.

#### cspann: add BestCentroids vector index test

Add a new "best-centroids" test that prints out the partitions with the
closest centroids for a query vector. This is useful when gauging the
quality of the index.

Also, update the "recall" test to only sample from vectors that are not
part of the index. Achieving high recall is more challenging when
searching for such vectors.

#### cspann: move level searcher init code to Init method

Previously, the searcher initialized the search set for each level in
the Next method. This commit moves that code into the levelSearcher.Init
method. This makes some planned refactoring easier in an another commit.

#### cspann: add utils.ReplaceWithLast helper function

Add generic ReplaceWithLast function that removes an element from a
slice by replacing it with the last element and truncating the slice.

#### vecindex: add TryMoveVector method to Store interface

Add a new TryMoveVector to the Store interface. This method moves a
single vector from one partition to another as an atomic operation,
provided the destination metadata is as expected.

andy-kimball committed:
vecindex: add TryMoveVector method to Store interface

Add a new TryMoveVector to the Store interface. This method moves a
single vector from one partition to another as an atomic operation,
provided the destination metadata is as expected.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
cspann: add utils.ReplaceWithLast helper function

Add generic ReplaceWithLast function that removes an element from a
slice by replacing it with the last element and truncating the slice.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
cspann: move level searcher init code to Init method

Previously, the searcher initialized the search set for each level in
the Next method. This commit moves that code into the levelSearcher.Init
method. This makes some planned refactoring easier in an another commit.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
cspann: add BestCentroids vector index test

Add a new "best-centroids" test that prints out the partitions with the
closest centroids for a query vector. This is useful when gauging the
quality of the index.

Also, update the "recall" test to only sample from vectors that are not
part of the index. Achieving high recall is more challenging when
searching for such vectors.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
cspann: enhance SearchSet

Add support in SearchSet for excluding partitions during search. Any
vectors in these partitions will not be added to the set. Also add
an option that includes the distance of each vector from its centroid
in search results. These options will be used by the split operation.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
cspann: move vectors from sibling partitions during split

When a partition is split, the left and right target partitions are assigned
new centroids. This makes it possible for vectors in other partitions at the
same level to now be closer to one of those new centroids than they are to
their own centroid. In that case, we need to move those vectors to whichever
target partition is now closer.

Epic: CRDB-42943
Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just a question and nit.

> Reviewed 6 of 6 files at r1, 11 of 11 files at r2, 2 of 2 files at r3, 4 of 4 files at r4, 4 of 4 files at r5, 1 of 16 files at r6, 20 of 20 files at r7, 4 of 4 files at r8, 16 of 16 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148713#-:-OUCCRKl15TxwiOmQJEB:b-i9rv7w)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/vecindex/cspann/testdata/insert.ddt` line 141 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148713#-OUCC9q11incvtNPp7av:-OUCC9q11incvtNPp7aw:b-sjq7ol) ([raw file](https://github.com/cockroachdb/cockroach/blob/497bb6d05d20140ca1341e844c38efdcc7ac68eb/pkg/sql/vecindex/cspann/testdata/insert.ddt#L141)):*
> ```
> • 1 (0, 0)
> │
> ├───• 8 (0.4, 0.8)
> ```

Isn't this more levels than necessary?
___
*[`pkg/sql/vecindex/cspann/index.go` line 615 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148713#-OUCA3t6Cy-NEGHG9Suu:-OUCA3t6Cy-NEGHG9Suv:b-zf210y) ([raw file](https://github.com/cockroachdb/cockroach/blob/497bb6d05d20140ca1341e844c38efdcc7ac68eb/pkg/sql/vecindex/cspann/index.go#L615)):*
> ```Go
> 		SkipRerank:   true,
> 		UpdateStats:  !vi.options.DisableAdaptiveSearch,
> 	}, LeafLevel, vec, false /* isRandomized */)
> ```

nit: these comments should say `transformed` instead of `isRandomized` or `randomized`.


<!-- Sent from Reviewable.io -->

mw5h reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 6 files at r1, 11 of 11 files at r2, 2 of 2 files at r3, 4 of 4 files at r4, 20 of 20 files at r7, 4 of 4 files at r8, 16 of 16 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148713#-:-OUHnSgx-3KW1yCWJPG_:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148713)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/vecindex/cspann/index.go` line 615 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148713#-OUCA3t6Cy-NEGHG9Suu:-OUKrkNG5gcZjVBGGm3i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/497bb6d05d20140ca1341e844c38efdcc7ac68eb/pkg/sql/vecindex/cspann/index.go#L615)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: these comments should say `transformed` instead of `isRandomized` or `randomized`.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/cspann/testdata/insert.ddt` line 141 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148713#-OUCC9q11incvtNPp7av:-OUKqOjY-ioy3nZcwi-1:b-2bn3n8) ([raw file](https://github.com/cockroachdb/cockroach/blob/497bb6d05d20140ca1341e844c38efdcc7ac68eb/pkg/sql/vecindex/cspann/testdata/insert.ddt#L141)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Isn't this more levels than necessary?
</blockquote></details>

A split can occur when # levels >= max partition size, which is 4 in this case. `searchForUpdateHelper` uses `>=` rather than `>` because it uses an inconsistent scan of the partition that can happen before a new vector is inserted. That is, we search for the partition to insert into and only later actually insert the vector. So there's not a sharp transactional boundary for a partition that is oversized - it's just an estimate.


<!-- Sent from Reviewable.io -->

andy-kimball committed:
cspann: do not duplicate vectors to avoid an empty partition

Previously, the split fixup would sometimes add an interior centroid
vector to two different parent partitions. In the past, this was
necessary in order to avoid creating an empty partition that could
unbalance the K-means tree. But this is no longer necessary; empty
partitions are now allowed. So this commit removes that code.

Epic: CRDB-42943
Release note: None
